File: octave/triangular.m
%
% Triangular matrix resolution method
%
% input:
%  - A: triangular matrix
%  - b: known terms vector
%  - type: 0 if A is upper triangular, 1 if is lower triangular
% output:
%  - x: the results vector
%
function x = triangular(A, b, type)
  n = length(b);
  if type == 1
    start = 1;
    stop  = n;
    step  = 1;
  else
    start = n;
    stop  = 1;
    step  = -1;
  end
  for i=start:step:stop
    x(i) = b(i);
    for j=start:step:i-step
      x(i) = x(i)-A(i,j)*x(j);
    end
    x(i) = x(i)/A(i,i);
  end
  x = x'; % return a column vector 
  return
end

File: python/triangular.py
# Triangular linear system solver

import numpy as np


def triangular(A, b, typ):
    '''
    Input:
      - A: triangular matrix
      - b: known terms vector
      - typ: 0 if upper, 1 if lower
    Output:
      - x: solution vector
    '''
    n = len(b)
    x = np.empty([n,1])
    if typ == 1:
        start = 0
        stop  = n
        step  = 1
    else:
        start = n-1
        stop  = -1
        step  = -1
    for i in range(start,stop,step):
        x[i] = b[i]
        for j in range(start,i,step):
            x[i] = x[i] - A[i,j]*x[j]
        x[i] = float(x[i]) / A[i,i]
    return x